hail/src/main/scala/is/hail/backend/spark/LowerTableIR.scala

```diff
@@ -1,29 +1,213 @@
+
+    wrapper +=
+      s"""
+         |try {
+         |  return (long) process_partition(((ScalaRegion *)${ wrapper.getArg(1) })->get_wrapped_region(), (char *)${ wrapper.getArg(2) }, (char *)${ wrapper.getArg(3) });
+         |} catch (const FatalError& e) {
+         |  NATIVE_ERROR(${ wrapper.getArg(0) }, 1005, e.what());
+         |  return -1;
+         |}
+       """.stripMargin
+    wrapper.end()
+    val mod = tub.end().build("-O2")
+    val st = new NativeStatus()
+    mod.findOrBuild(st)
+    assert(st.ok, st.toString())
+
+    val key = mod.getKey
+    val bin = mod.getBinary
+
+    val localContextType = contextType
 
-class LowerTableIR {
+    val resultEncoder = codecSpec.buildEncoder(resultType)
+
+    val values = sc.parallelize(context, context.length).mapPartitions { ctxIt =>
+      val ctx = ctxIt.next()
+      val st = new NativeStatus()
+
+      val mod = new NativeModule(key, bin)
+      val f = mod.findLongFuncL3(st, "process_partition_wrapper")
+      assert(st.ok, st.toString())
+
+      Region.scoped { region =>
+        val rvb = new RegionValueBuilder(region)
+
+        val globalsOff = RegionValue.fromBytes(gDecoder, region, globsBC)
+
+        rvb.start(localContextType.physicalType)
+        rvb.addAnnotation(localContextType, ctx)
+        val contextOff = rvb.end()
+
+        val result = f(st, region.get(), globalsOff, contextOff)
+        if (st.fail)
+          fatal(st.toString())
+
+        val encoded = RegionValue.toBytes(resultEncoder, region, result)
+        Iterator.single(encoded)
+      }
+    }.collect()
+
+    rvb.startArray(values.length)
+    val dec = codecSpec.buildDecoder(resultType, resultType)
+    values.foreach { v =>
+      val off = RegionValue.fromBytes(dec, region, v)
+      if (resultType.isFieldDefined(region, off, 0))
+        rvb.addRegionValue(resultType.types(0), region, resultType.loadField(region, off, 0))
+      else
+        rvb.setMissing()
+    }
+    rvb.endArray()
+  }
+}
+
+object LowerTableIR {
+
+  def lower(ir: IR): SparkPipeline = ir match {
+
+    case TableCount(tableIR) =>
+      val stage = lower(tableIR)
+      val counts = Ref(genUID(), TArray(TInt64()))
+      SparkPipeline(
+        Map(counts.name -> stage.copy(body = Cast(ArrayLen(stage.body), TInt64()))),
+        invoke("sum", counts))
+
+    case TableGetGlobals(child) =>
+      lower(child).globals.head.value
+
+    case TableCollect(child) =>
+      val lowered = lower(child)
+      val globals = lowered.globals.head
+      val rows = Ref(genUID(), TArray(lowered.body.typ))
+      assert(lowered.body.typ.isInstanceOf[TContainer])
+      val elt = genUID()
+      SparkPipeline(
+        Map((rows.name, lowered) +: globals.value.stages.toSeq: _*),
+        MakeStruct(FastIndexedSeq(
+          "rows" -> ArrayFlatMap(rows, elt, Ref(elt, lowered.body.typ)),
+          "global" -> lowered.globals.head.value.body)))
```

Comment:
`lowered.globals.head` appears in a couple places. Why is it safe to ignore the rest of it? What is `lowered.globals.head`? what are the tail?

Reply:
`globals.head` is the globals of the current table. the rest of it is basically any other global information that we might want to include (e.g. the globals of the right table in a left-joined table). I thought it was documented, but maybe not--i can leave a note.